File: accommodation-booking/server/src/Services/Identity/Identity.API/Integration/SubscriptionManager/SubscriptionManager.cs
namespace Identity.API.Integration.SubscriptionManager
{
    public class SubscriptionManager : ISubscriptionManager
    {
        private readonly Dictionary<string, Type> _handlers;
        private readonly List<Type> _eventTypes;

        public SubscriptionManager()
        {
            _handlers = new Dictionary<string, Type>();
            _eventTypes = new List<Type>();
        }

        public void AddSubcription<T, TH>()
            where T : IntegrationEvent
            where TH : IIntegrationEventHandler<T>
        {
            var eventName = typeof(T).Name;

            if (HasSubscriptionsForEvent(eventName))
                return;

            _handlers.Add(eventName, typeof(TH));
            _eventTypes.Add(typeof(T));
        }

        public Type GetHandlerForEvent(string eventName) => _handlers[eventName];

        public bool HasSubscriptionsForEvent(string eventName) => _handlers.ContainsKey(eventName);

        public Type GetEventTypeByName(string eventName) => _eventTypes.SingleOrDefault(t => t.Name == eventName);

    }
}

File: accommodation-booking/server/src/Services/Accomodation/src/main/java/accomodation/integration/subscriptionManager/SubscriptionManager.java
package accomodation.integration.subscriptionManager;

import accomodation.integration.IIntegrationEventHandler;
import accomodation.integration.IntegrationEvent;
import org.springframework.stereotype.Service;

import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;

@Service
public class SubscriptionManager implements ISubscriptionManager {

    private Map<String, Class> handlers;
    private Map<String, Class> eventTypes;

    public SubscriptionManager() {
        this.handlers = new HashMap<>();
        this.eventTypes = new HashMap<>();
    }

    @Override
    public <T extends IntegrationEvent, TH extends IIntegrationEventHandler> void addSubscription(Class<T> eventType, Class<TH> eventHandlerType) {
        handlers.put(eventType.getSimpleName(), eventHandlerType);
        eventTypes.put(eventType.getSimpleName(), eventType);
    }

    @Override
    public boolean hasSubscriptionsForEvent(String eventName) {
        return false;
    }

    @Override
    public Class getHandlerForEvent(String eventName) { return this.handlers.get(eventName); }

    @Override
    public Class getEventTypeByName(String eventName) {
        return this.eventTypes.get(eventName);
    }
}
